internal/packer_registry/types.bucket.go

```diff
@@ -0,0 +1,387 @@
+					// Iteration's expectedBuilds so we don't overwrite them.
+					b.Iteration.expectedBuilds = append(b.Iteration.expectedBuilds[:i],
+						b.Iteration.expectedBuilds[i+1:]...)
+					log.Printf("build of component type %s is already marked DONE; "+
+						"removing the build from the HCP Packer Registry expected builds.", expected)
+				}
+				break
+			}
+		}
+		if !found {
+			toCreate = append(toCreate, existing.ComponentType)
+		}
+	}
+
+	for _, buildName := range toCreate {
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			log.Printf("[TRACE] registering build with iteration for %q.", name)
+			// Need a way to handle skipping builds that were already created.
+			// TODO when we load an existing iteration we will probably have a build Id so we should skip.
+			// we also need to bubble up the errors here.
+			err := b.CreateInitialBuildForIteration(ctx, name)
+			if checkErrorCode(err, codes.AlreadyExists) {
+				// Check whether build is complete, and if so, skip it.
+				log.Printf("[TRACE] build %s already exists in PAR, continuing...")
+			} else {
+				errs = multierror.Append(errs, err)
+			}
+		}(buildName)
+	}
+	wg.Wait()
+
+	return errs.ErrorOrNil()
+}
+
+// connect initializes a client connection to a remote HCP Packer Registry service on HCP.
+// Upon a successful connection the initialized client is persisted on the Bucket b for later usage.
+func (b *Bucket) connect() error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
+
+	registryClient, err := NewClient(b.Config)
+	if err != nil {
+		return errors.New("Failed to create client connection to artifact registry: " + err.Error())
+	}
+	b.client = registryClient
+	return nil
+}
+
+func (b *Bucket) RegisterBuildForComponent(sourceName string) {
+	if b == nil {
+		return
+	}
+
+	if _, ok := b.Iteration.builds.Load(sourceName); ok {
+		return
+	}
+	b.Iteration.expectedBuilds = append(b.Iteration.expectedBuilds, sourceName)
+}
+
+func (b *Bucket) PublishBuildStatus(ctx context.Context, name string, status models.HashicorpCloudPackerBuildStatus) error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
+
+	// Lets check if we have something already for this build
+	build, ok := b.Iteration.builds.Load(name)
+	if !ok {
+		return fmt.Errorf("no build for the component %q associated to the iteration %q", name, b.Iteration.ID)
+	}
+
+	buildToUpdate, ok := build.(*Build)
+	if !ok {
+		return fmt.Errorf("the build for the component %q does not appear to be a valid registry Build", name)
+	}
+
+	if buildToUpdate.ID == "" {
+		return fmt.Errorf("the build for the component %q does not have a valid id", name)
+	}
+
+	buildInput := &models.HashicorpCloudPackerUpdateBuildRequest{
+		BuildID: buildToUpdate.ID,
+		Updates: &models.HashicorpCloudPackerBuildUpdates{
+			PackerRunUUID: buildToUpdate.RunUUID,
+			Labels:        buildToUpdate.Metadata,
+			Status:        &status,
+		},
+	}
+
+	// Possible bug of being able to set DONE with no RunUUID being set.
```

Comment:
This comment didn't help me a lot 👀 . Is this a TODO fix?

Reply:
Yes it is a TODO.